aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/ApplicationLifeCycleHelper.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.analytics.pinpoint;
+
+import android.app.Activity;
+import android.app.Application;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.os.Bundle;
+import android.util.Log;
+
+import com.amazonaws.mobile.auth.core.internal.util.ThreadUtils;
+
+import java.util.WeakHashMap;
+
+/**
+ * Aids in determining when your application has entered or left the foreground.
+ * The constructor registers to receive Activity lifecycle events and also registers a
+ * broadcast receiver to handle the screen being turned off.  Abstract methods are
+ * provided to handle when the application enters the background or foreground.
+ * Any activity lifecycle callbacks can easily be overriden if additional handling
+ * is needed. Just be sure to call through to the super method so that this class
+ * will still behave as intended.
+ **/
+public abstract class ApplicationLifeCycleHelper implements Application.ActivityLifecycleCallbacks {
+    private static final String LOG_TAG = ApplicationLifeCycleHelper.class.getSimpleName();
+    private static final String ACTION_SCREEN_OFF = "android.intent.action.SCREEN_OFF";
+    private boolean inForeground = false;
+    /** Tracks the lifecycle of activities that have not stopped (including those restarted). */
+    private WeakHashMap<Activity, String> activityLifecycleStateMap = new WeakHashMap<>();
```

Comment:
`ActivityBase` and `AppCompatActivity` have existing solutions to track/obtain the current lifecycle. Checkout the AndroidX Lifecycle Components, with which V7 App Compat Activity is integrated.

However, I can see an advantage to having something that works with every single Activity, ever. And for that, I guess you might need to have your own tracker.

A good name for this class might be `ApplicationForegroundTracker`, instead of `ApplicationLifecycleHelper`.

Anyway, even if you implement your own tracker, you can still use as much of AndroidX Lifecycle here as possible. Instead of using strings, you should use `enum`s, which provide a bounded set of possible states the map value can be in. And, as it turns out, that `enum` is already defined by AndroidX Lifecycle --

So this map would (probably) become:
```
private Map<Activity, Lifecycle.State> activityStates = new WeakHashMap<>();
```
(Ref: [`LifecycleState`](https://developer.android.com/reference/android/arch/lifecycle/Lifecycle.State))

Reply:
I have updated to use LifecycleState and updated the naming as well.